### PR TITLE
Fix selection of nodes for cinder-volume, ceilometer-agent and nova-compute

### PR DESCRIPTION
**Pull request by rhafer**
Commit 45a66ee90 broke HA deployments on releases other than Cloud 7
that had more than one possible node for cinder-volume, ceilometer-agent
or nova-compute. This is caused by not breaking $unclustered_nodes into
an array as expected by the rest of the code in custom_configuration().

Additionally, we add some quoting for the -z tests to avoid whitespace
issues that triggered "[: too many arguments" errors in the past.

### CODE REVIEW

**Comment by rhafer** [comment]
This is a more complete version of https://github.com/SUSE-Cloud/automation/pull/1225

**Comment by abel-navarro** [comment]
looks good, but let me test it on an actual deployment (about 1 hour).

**Comment by rhafer** [comment]
> looks good, but let me test it on an actual deployment (about 1 hour).

@abel-navarro I've also added a couple of more ci jobs here to test it at least for HA deployments on Cloud 6 and 7.

**Comment by abel-navarro** [comment]
my deployments have not finished yet, but at least they have gone past the point of the ceph proposal apply. I'm trying with cloud6-ha with nodenumber=4 and nodenumber=5.

**Comment by abel-navarro** [comment]
+1

File: scripts/qa_crowbarsetup.sh [hunk 1]
```diff
@@ -2848,14 +2848,14 @@ function custom_configuration()
...
-                local novanodes=$unclustered_nodes
+                local novanodes=($unclustered_nodes)
 
                 # make sure we do not pick SP1 nodes on cloud7
                 if [ -n "$deployceph" ] && iscloudver 7 ; then
                     novanodes=$unclustered_sles12plusnodes
                 fi
 
-                if [ -z $novanodes ]; then
+                if [ -z "$novanodes" ]; then
```
Comment (aspiers) [inline scripts/qa_crowbarsetup.sh]: As per #962 I guess @nicolasbock and @jdsn would argue this should be `[[ $novanodes ]]` instead :-)

File: scripts/qa_crowbarsetup.sh [hunk 2]
```diff
@@ -2907,12 +2907,12 @@ function custom_configuration()
...
                 proposal_set_value ceilometer default "['attributes']['ceilometer']['use_mongodb']" "false"
 
-                local ceilometernodes=$unclustered_nodes
+                local ceilometernodes=($unclustered_nodes)
                 # make sure we do not pick SP1 nodes on cloud7
                 if [ -n "$deployceph" ] && iscloudver 7 ; then
                     ceilometernodes=$unclustered_sles12plusnodes
                 fi
-                if [ -z $ceilometernodes ]; then
+                if [ -z "$ceilometernodes" ]; then
```
Comment (aspiers) [inline scripts/qa_crowbarsetup.sh]: Same here.

File: scripts/qa_crowbarsetup.sh [hunk 3]
```diff
@@ -3022,12 +3022,12 @@ function custom_configuration()
...
             if [[ $hacloud = 1 ]] ; then
                 # fetch one of the compute nodes as cinder_volume
-                local cinder_volume=$unclustered_nodes
+                local cinder_volume=($unclustered_nodes)
                 # make sure we do not pick SP1 nodes on cloud7
                 if [ -n "$deployceph" ] && iscloudver 7 ; then
                     cinder_volume=$unclustered_sles12plusnodes
                 fi
-                if [ -z $cinder_volume ]; then
+                if [ -z "$cinder_volume" ]; then
```
Comment (aspiers) [inline scripts/qa_crowbarsetup.sh]: and here (just testing out github's new PR comment bundling ;-) )

File: scripts/qa_crowbarsetup.sh (hunk 1 above)
Comment (nicolasbock) [inline scripts/qa_crowbarsetup.sh]: @aspiers How did you know :)

**Comment by abel-navarro** [comment]
With `nodenumber=4` it deploys properly. The cinder-volume is assigned to **compute1** which I think is as intended.

image

image


I did not properly configure the deployment with `nodenumber=5`, so I'm trying that again.

**Comment by aspiers** [comment]
`cct` failing on the `gating-ha-6` gate due to a controller being in `unknown` state.

**Comment by bmwiedemann** [comment]
My main concern is code duplication which means we always have to fix 3 places instead of one... but that can be addresses in a separate PR.

**Comment by abel-navarro** [comment]
I run again the deployment with `nodenumber=5`. This time also completed successfully but again the **cinder-volume**  node went to **data2**, which is the second controller in the HA cluster. It is not selecting the extra node for the volume.

I did not run `testsetup` so cannot say about `cct`.

**Comment by rhafer** [comment]
> I run again the deployment with nodenumber=5. This time also completed successfully but again the cinder-volume node went to data2

Hm, strange. Which ```cloudsource``` was that for?

File: scripts/qa_crowbarsetup.sh (hunk 1 above)
Comment (rhafer) [inline scripts/qa_crowbarsetup.sh]: Done. I guess we wouldn't strictly need the quotes now anymore but I left them in place.

**Comment by rhafer** [comment]
> I run again the deployment with nodenumber=5. This time also completed successfully but again the cinder-volume node went to data2, which is the second controller in the HA cluster. It is not selecting the extra node for the volume.

@abel-navarro Hm, I am not able to reproduce that. In my testrun (develcloud6, ha enable) it picked the first compute node. Which parameters did you use for building?

**Comment by abel-navarro** [comment]
@rhafer deployments went fine for me, and also `testsetup` and `cct`. These are the params for the builds:

```
export cephvolumenumber=1
export want_neutronsles12=1
export want_mtu_size=8900

export cloudsource="develcloud6"
export mkcloudtarget="all_noreboot"
export libvirt_type="kvm"
export networkingplugin="openvswitch"
export networkingmode="vxlan"
export storage_method="default"
export TESTHEAD="1"
export nodenumber="4"
export tempestoptions="-t -s"
export hacloud="1"
export clusterconfig="data+services+network=2"
export label="openstack-mkcloud-SLE12-x86_64"
export vcpus="2"
```

also tried with `nodenumber="5"`

**Comment by rhafer** [comment]
Hm, the cloud7 job currently fails because chef-server doesn't start on the admin node. (So unrelated to this PR).

**Comment by abel-navarro** [comment]
In the re-deploys I focused on the tests not on the `cinder-volume` assignment. After checking again, `cinder-volume` goes always to the first compute node.

**4-node**
image

**5-node**
image
